### PR TITLE
Improve performance of RBAC model

### LINKED ISSUE #87: Improve performance of RBAC model
RABC model has poorer performance than other models from the result of the benchmark. I will try to improve it.

### PR DESCRIPTION
**Pull request by Sagilio**
close #87 and split to #93

### CODE REVIEW

**Comment by hsluoyz** [comment]
@huazhikui @xcaptain @nodece please review.

**Review by xcaptain — APPROVED** [review]
So much code changed, don't know why these changes improve performance, but since tests passed, so it should be right. Would take a deeper look later.

**Comment by hsluoyz** [comment]
@Sagilio can you describe how it is improved?

**Comment by nodece** [comment]
@Sagilio Your did a great job, could you splitting multiple PR submissions? there are too many file changes here.

**Comment by Sagilio** [comment]
> @Sagilio Your did a great job, could you splitting multiple PR submissions? there are too many file changes here.

I will try to split into two PRs;

**Comment by Sagilio** [comment]
> @Sagilio can you describe how it is improved?

Ok, I will describe the reason and result for these changes later.

**Review by huazhikui — APPROVED** [review]

File: NetCasbin/InternalEnforcer.cs [hunk 1]
```diff
@@ -387,6 +387,7 @@ private void NoticePolicyChanged(string section)
             if (autoBuildRoleLinks && section.Equals(PermConstants.Section.RoleSection))
             {
                 BuildRoleLinks();
+                ExpressionProvider.ClearFunctions();
```
Comment (nodece) [inline NetCasbin/InternalEnforcer.cs:390]: unnecessary.

File: NetCasbin/InternalEnforcer.cs [hunk 2]
```diff
@@ -400,6 +401,7 @@ private async Task NoticePolicyChangedAsync(string section)
...
```
Comment (nodece) [inline NetCasbin/InternalEnforcer.cs:404]: unnecessary.

File: NetCasbin/InternalEnforcer.cs (hunk 1 above)
Comment (Sagilio) [reply NetCasbin/InternalEnforcer.cs:390]: The cache of GFunction needs to refresh. If delete it the unit test can not pass.

Comment (nodece) [reply NetCasbin/InternalEnforcer.cs:390]: but it will clear user add function.

Comment (Sagilio) [reply NetCasbin/InternalEnforcer.cs]: You are right, I have recommitted to solve it.

File: NetCasbin/InternalEnforcer.cs (hunk 2 above)
Comment (Sagilio) [reply NetCasbin/InternalEnforcer.cs]: like above.

**Pull request merged**